Fix assertion error in instructorCourseJoinEmail on live server

v6.1
Found many of these in log.
On a related note, these are logged as 'warning' rather than 'severe'. Shouldn't they be severe?
```
com.google.apphosting.runtime.jetty9.JettyLogger warn: Error for /worker/instructorCourseJoinEmail (JettyLogger.java:29)
java.lang.AssertionError: null
	at teammates.common.util.Assumption.fail(Assumption.java:61)
	at teammates.common.util.Assumption.assertTrue(Assumption.java:29)
	at teammates.common.util.Assumption.assertNotNull(Assumption.java:272)
	at teammates.common.util.Assumption.assertNotNull(Assumption.java:263)
	at teammates.ui.automated.InstructorCourseJoinEmailWorkerAction.execute(InstructorCourseJoinEmailWorkerAction.java:37)
	at teammates.ui.automated.AutomatedServlet.doPost(AutomatedServlet.java:41)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:848)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1772)
	at com.googlecode.objectify.ObjectifyFilter.doFilter(ObjectifyFilter.java:48)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.runtime.jetty9.SaveSessionFilter.doFilter(SaveSessionFilter.java:37)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:48)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:582)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
	at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at com.google.apphosting.runtime.jetty9.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:297)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
	at org.eclipse.jetty.server.Server.handle(Server.java:534)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320)
	at com.google.apphosting.runtime.jetty9.RpcConnection.handle(RpcConnection.java:202)
	at com.google.apphosting.runtime.jetty9.RpcConnector.serviceRequest(RpcConnector.java:81)
	at com.google.apphosting.runtime.jetty9.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:108)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:680)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:642)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:612)
	at com.google.apphosting.runtime.JavaRuntime$NullSandboxRequestRunnable.run(JavaRuntime.java:806)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:274)
	at java.lang.Thread.run(Thread.java:745)
```

Got a few today 😢 :
>Error Message: teammates.common.util.Logger severe: Exception occured while performing http://teammatesv4.appspot.com/worker/instructorCourseJoinEmail|||
java.lang.AssertionError: null
at teammates.common.util.Assumption.fail(Assumption.java:61)
at teammates.common.util.Assumption.assertTrue(Assumption.java:29)
at teammates.common.util.Assumption.assertNotNull(Assumption.java:272)
at teammates.common.util.Assumption.assertNotNull(Assumption.java:263)
at teammates.ui.automated.InstructorCourseJoinEmailWorkerAction.execute(InstructorCourseJoinEmailWorkerAction.java:48)
at teammates.ui.automated.AutomatedServlet.doPost(AutomatedServlet.java:46)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:848)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1772)
at com.googlecode.objectify.ObjectifyFilter.doFilter(ObjectifyFilter.java:48)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1759)
at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:582)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)
at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
at com.google.apphosting.runtime.jetty9.ParseBlobUploadHandler.handle(ParseBlobUploadHandler.java:120)
at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1182)
at com.google.apphosting.runtime.jetty9.AppEngineWebAppContext.doHandle(AppEngineWebAppContext.java:171)
at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512)
at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112)
at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
at com.google.apphosting.runtime.jetty9.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:297)
at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134)
at org.eclipse.jetty.server.Server.handle(Server.java:534)
at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:320)
at com.google.apphosting.runtime.jetty9.RpcConnection.handle(RpcConnection.java:202)
at com.google.apphosting.runtime.jetty9.RpcConnector.serviceRequest(RpcConnector.java:81)
at com.google.apphosting.runtime.jetty9.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:123)
at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchServletRequest(JavaRuntime.java:693)
at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.dispatchRequest(JavaRuntime.java:655)
at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:625)
at com.google.apphosting.runtime.JavaRuntime$NullSandboxRequestRunnable.run(JavaRuntime.java:819)
at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:274)
at java.lang.Thread.run(Thread.java:745)

--- a/src/main/java/teammates/ui/automated/InstructorCourseJoinEmailWorkerAction.java
+++ b/src/main/java/teammates/ui/automated/InstructorCourseJoinEmailWorkerAction.java
@@ -47,7 +47,7 @@ public class InstructorCourseJoinEmailWorkerAction extends AutomatedAction {
         // saving the new instructor takes more time in GAE.
         // The instructor `id` can be constructed back using (instructorEmail%courseId)
         // because instructors' email cannot be changed before joining the course.
-        RetryManager rm = new RetryManager(5);
+        RetryManager rm = new RetryManager(4);
         InstructorAttributes instructor = null;
 
         try {

--- a/src/main/java/teammates/ui/automated/InstructorCourseJoinEmailWorkerAction.java
+++ b/src/main/java/teammates/ui/automated/InstructorCourseJoinEmailWorkerAction.java
@@ -7,6 +7,9 @@ import teammates.common.exception.TeammatesException;
 import teammates.common.util.Assumption;
 import teammates.common.util.Const.ParamsNames;
 import teammates.common.util.EmailWrapper;
+import teammates.common.util.retry.MaximumRetriesExceededException;
+import teammates.common.util.retry.RetryManager;
+import teammates.common.util.retry.RetryableTaskReturns;
 import teammates.logic.api.EmailGenerator;
 
 /**
@@ -44,7 +47,21 @@ public class InstructorCourseJoinEmailWorkerAction extends AutomatedAction {
         // saving the new instructor takes more time in GAE.
         // The instructor `id` can be constructed back using (instructorEmail%courseId)
         // because instructors' email cannot be changed before joining the course.
-        InstructorAttributes instructor = logic.getInstructorById(courseId, instructorEmail);
+        RetryManager rm = new RetryManager(5);
+        InstructorAttributes instructor = null;
+
+        try {
+            instructor = rm.runUntilNotNull(new RetryableTaskReturns<InstructorAttributes>("getInstructor") {
+                @Override
+                public InstructorAttributes run() {
+                    return logic.getInstructorById(courseId, instructorEmail);
+                }
+            });
+        } catch (MaximumRetriesExceededException e) {
+            // This Assumption fail ensures that the task is retried
+            Assumption.fail("If the instructor hasn't been retrieved yet there is some problem in adding of instructor"
+                    + TeammatesException.toStringWithStackTrace(e));
+        }
         Assumption.assertNotNull(instructor);
 
         EmailWrapper email = new EmailGenerator()

--- a/src/main/java/teammates/ui/automated/StudentCourseJoinEmailWorkerAction.java
+++ b/src/main/java/teammates/ui/automated/StudentCourseJoinEmailWorkerAction.java
@@ -6,6 +6,9 @@ import teammates.common.exception.TeammatesException;
 import teammates.common.util.Assumption;
 import teammates.common.util.Const.ParamsNames;
 import teammates.common.util.EmailWrapper;
+import teammates.common.util.retry.MaximumRetriesExceededException;
+import teammates.common.util.retry.RetryManager;
+import teammates.common.util.retry.RetryableTaskReturns;
 import teammates.logic.api.EmailGenerator;
 
 /**
@@ -35,7 +38,21 @@ public class StudentCourseJoinEmailWorkerAction extends AutomatedAction {
 
         CourseAttributes course = logic.getCourse(courseId);
         Assumption.assertNotNull(course);
-        StudentAttributes student = logic.getStudentForEmail(courseId, studentEmail);
+
+        RetryManager rm = new RetryManager(5);
+        StudentAttributes student = null;
+
+        try {
+            student = rm.runUntilNotNull(new RetryableTaskReturns<StudentAttributes>("getStudent") {
+                @Override
+                public StudentAttributes run() throws RuntimeException {
+                    return logic.getStudentForEmail(courseId, studentEmail);
+                }
+            });
+        } catch (MaximumRetriesExceededException e) {
+            Assumption.fail("If the student hasn't been retrieved yet there is some problem in adding of student"
+                    + TeammatesException.toStringWithStackTrace(e));
+        }
         Assumption.assertNotNull(student);
 
         EmailWrapper email = isRejoin

--- a/src/main/java/teammates/ui/automated/StudentCourseJoinEmailWorkerAction.java
+++ b/src/main/java/teammates/ui/automated/StudentCourseJoinEmailWorkerAction.java
@@ -39,7 +39,7 @@ public class StudentCourseJoinEmailWorkerAction extends AutomatedAction {
         CourseAttributes course = logic.getCourse(courseId);
         Assumption.assertNotNull(course);
 
-        RetryManager rm = new RetryManager(5);
+        RetryManager rm = new RetryManager(4);
         StudentAttributes student = null;
 
         try {

--- a/src/main/webapp/WEB-INF/queue.xml
+++ b/src/main/webapp/WEB-INF/queue.xml
@@ -102,10 +102,10 @@
         <rate>5/s</rate>
         <bucket-size>20</bucket-size>
         <retry-parameters>
-            <task-retry-limit>5</task-retry-limit>
-            <min-backoff-seconds>2</min-backoff-seconds>
-            <max-backoff-seconds>4</max-backoff-seconds>
-            <max-doublings>0</max-doublings>
+            <task-retry-limit>3</task-retry-limit>
+            <min-backoff-seconds>5</min-backoff-seconds>
+            <max-backoff-seconds>40</max-backoff-seconds>
+            <max-doublings>2</max-doublings>
         </retry-parameters>
     </queue>
 
@@ -129,10 +129,10 @@
         <rate>5/s</rate>
         <bucket-size>20</bucket-size>
         <retry-parameters>
-            <task-retry-limit>5</task-retry-limit>
-            <min-backoff-seconds>2</min-backoff-seconds>
-            <max-backoff-seconds>4</max-backoff-seconds>
-            <max-doublings>0</max-doublings>
+            <task-retry-limit>3</task-retry-limit>
+            <min-backoff-seconds>5</min-backoff-seconds>
+            <max-backoff-seconds>40</max-backoff-seconds>
+            <max-doublings>2</max-doublings>
         </retry-parameters>
     </queue>
 

--- a/src/main/webapp/WEB-INF/queue.xml
+++ b/src/main/webapp/WEB-INF/queue.xml
@@ -102,7 +102,10 @@
         <rate>5/s</rate>
         <bucket-size>20</bucket-size>
         <retry-parameters>
-            <task-retry-limit>2</task-retry-limit>
+            <task-retry-limit>5</task-retry-limit>
+            <min-backoff-seconds>2</min-backoff-seconds>
+            <max-backoff-seconds>4</max-backoff-seconds>
+            <max-doublings>0</max-doublings>
         </retry-parameters>
     </queue>
 
@@ -126,7 +129,10 @@
         <rate>5/s</rate>
         <bucket-size>20</bucket-size>
         <retry-parameters>
-            <task-retry-limit>2</task-retry-limit>
+            <task-retry-limit>5</task-retry-limit>
+            <min-backoff-seconds>2</min-backoff-seconds>
+            <max-backoff-seconds>4</max-backoff-seconds>
+            <max-doublings>0</max-doublings>
         </retry-parameters>
     </queue>
 

